module/zfs/ddt.c

```diff
@@ -719,6 +732,13 @@ ddt_lookup(ddt_t *ddt, const blkptr_t *bp, boolean_t add)
 
 	ddt_enter(ddt);
 
+	if (error == ENOENT && add == B_FALSE) {
+		avl_remove(&ddt->ddt_tree, dde);
```

Comment:
It would trigger an assert due to dde->dde_loading/dde_loaded in that case.

So technically, I don't believe so, but the real answer is that yes, there _may_ be a race condition there.

Reply:
Might the best approach be to actually defer the `avl_insert()` (currently like 706) and remove the `avl_remove()` from like 736, and put the `avl_insert()` at line 750? (We'd need to have a flag to know when we need to add it, vs it is already on the avl list), but that would ensure we don't end up with it getting flushed to disk when we `ddt_exit()` to search the the DDT.

Reply:
Doesn't that just keep a race condition around?  Some other thread may end up doing the same thing with the same entry, no?

Reply:
Can I conclude that the only reason for the ddt_exit/ddt_enter around the loop on line 721 is because it may involve I/O and therefore may take a long time?